// flooks/dist/flooks.cjs.js
'use strict';

var react = require('react');

var __DEV__ = process.env.NODE_ENV !== 'production';

var ERR_MODEL = 'model should be a function';
var ERR_PAYLOAD = 'payload should be an object or a function';
var ERR_OUT_MODEL = 'model passed to get() is not initialized';
var emptyObj = {};

var noop = function noop() {
  return function () {
    return null;
  };
};

var notObj = function notObj(val) {
  return Object.prototype.toString.call(val) !== '[object Object]';
};

var map = new WeakMap();

var useModel = function useModel(model) {
  if (__DEV__ && typeof model !== 'function') throw new Error(ERR_MODEL);
  var localModel = react.useRef(emptyObj);

  var _useState = react.useState(false),
      setState = _useState[1];

  var onEffect = react.useRef(noop);
  react.useMemo(function () {
    var _ref = map.get(model) || [],
        globalModel = _ref[0],
        globalSubs = _ref[1]; // global model


    if (!globalModel) {
      var get = function get(outModel) {
        var _map$get;

        if (typeof outModel === 'undefined') return globalModel;
        var outData = (_map$get = map.get(outModel)) == null ? void 0 : _map$get[0];
        if (__DEV__ && !outData) throw new Error(ERR_OUT_MODEL);
        return outData;
      };

      var set = function set(payload) {
        if (typeof payload === 'function') {
          payload = payload(globalModel);
        } else if (__DEV__ && notObj(payload)) {
          throw new Error(ERR_PAYLOAD);
        }

        Object.assign(globalModel, payload);
        globalSubs.forEach(function (updater) {
          return updater(payload);
        });
      };

      globalModel = model({
        get: get,
        set: set
      });
      globalSubs = [];
      map.set(model, [globalModel, globalSubs]);
    } // local model


    var hasState = false;
    var hasUpdate = false;
    var localTarget = {};
    var localHandler = {
      get: function get(target, key) {
        var val = globalModel[key];

        if (typeof val !== 'function') {
          hasState = true;
          target[key] = val;
          return target[key];
        }

        target[key] = new Proxy(val, {
          get: function get(fn, fnKey) {
            if (fnKey === 'loading' && !('loading' in fn)) fn.loading = false;
            return fn[fnKey];
          },
          apply: function apply(fn, _this, args) {
            var res = fn.apply(void 0, args);

            if (!('loading' in fn) || !res || typeof res.then !== 'function') {
              target[key] = fn;
              return res;
            }

            var setLoading = function setLoading(loading) {
              target[key].loading = loading;
              setState(function (s) {
                return !s;
              });
            };

            target[key] = function () {
              var newRes = fn.apply(void 0, arguments);
              setLoading(true);
              return newRes.finally(function () {
                return setLoading(false);
              });
            };

            setLoading(true);
            return res.finally(function () {
              return setLoading(false);
            });
          }
        });
        return target[key];
      },
      set: function set(target, key, val) {
        if (key in target) {
          target[key] = val;
          hasUpdate = true;
        }

        return true;
      }
    };
    localModel.current = new Proxy(localTarget, localHandler);

    var localUpdater = function localUpdater(payload) {
      Object.assign(localModel.current, payload);

      if (hasUpdate) {
        setState(function (s) {
          return !s;
        });
        hasUpdate = false;
      }
    };

    globalSubs.push(localUpdater);

    onEffect.current = function () {
      if (hasState) {
        localHandler.get = function (target, key) {
          return target[key];
        };

        return function () {
          return globalSubs.splice(globalSubs.indexOf(localUpdater), 1);
        };
      }

      globalSubs.splice(globalSubs.indexOf(localUpdater), 1);
      localModel.current = localTarget;
    };
  }, [model]);
  react.useEffect(function () {
    return onEffect.current();
  }, []);
  return localModel.current;
};

module.exports = useModel;
